tests/unit/datastore_test.py

```diff
@@ -2,9 +2,195 @@
+                                    Rule,
+                                    Profile,
+                                    Rules,
+                                    Endpoint,
+                                    NoEndpointForContainer)
 from etcd import Client as EtcdClient
+from etcd import EtcdResult
+from netaddr import IPNetwork, IPAddress
+import json
+
+TEST_HOST_PATH = "/calico/host/TEST_HOST"
+IPV4_POOLS_PATH = "/calico/ipam/v4/pool/"
+IPV6_POOLS_PATH = "/calico/ipam/v6/pool/"
+TEST_PROFILE_PATH = "/calico/policy/profile/TEST/"
+ALL_PROFILES_PATH = "/calico/policy/profile/"
+ALL_ENDPOINTS_PATH = "/calico/host/"
+ALL_HOSTS_PATH = "/calico/host/"
+TEST_ENDPOINT_PATH = "/calico/host/TEST_HOST/workload/docker/1234/endpoint/" \
+                     "1234567890ab"
+TEST_CONT_ENDPOINT_PATH = "/calico/host/TEST_HOST/workload/docker/1234/" \
+                          "endpoint/"
+TEST_CONT_PATH = "/calico/host/TEST_HOST/workload/docker/1234/"
+CONFIG_PATH = "/calico/config/"
+
+# 4 endpoints, with 2 TEST profile and 2 UNIT profile.
+EP_56 = Endpoint("567890abcdef", "active", "11-22-33-44-55-66")
+EP_56.profile_id = "TEST"
+EP_78 = Endpoint("7890abcdef12", "active", "11-22-33-44-55-66")
+EP_78.profile_id = "TEST"
+EP_90 = Endpoint("90abcdef1234", "active", "11-22-33-44-55-66")
+EP_90.profile_id = "UNIT"
+EP_12 = Endpoint("1234567890ab", "active", "11-22-33-44-55-66")
+EP_12.profile_id = "UNIT"
+
+
+class TestRule(unittest.TestCase):
+
+    def test_create(self):
+        """
+        Test creating a rule from constructor.
+        """
+        rule1 = Rule(action="allow",
+                     src_tag="TEST",
+                     src_ports=[300, 400])
+        self.assertDictEqual({"action": "allow",
+                              "src_tag": "TEST",
+                              "src_ports": [300, 400]}, rule1)
+
+    def test_to_json(self):
+        """
+        Test to_json() method.
+        """
+        rule1 = Rule(action="deny",
+                     dst_net=IPNetwork("192.168.13.0/24"))
+        json_str = rule1.to_json()
+        expected = json.dumps({"action": "deny",
+                               "dst_net": "192.168.13.0/24"})
+        self.assertEqual(json_str, expected)
+
+        rule2 = Rule(action="deny",
+                     src_net="192.168.13.0/24")
+        json_str = rule2.to_json()
+        expected = json.dumps({"action": "deny",
+                               "src_net": "192.168.13.0/24"})
+        self.assertEqual(json_str, expected)
+
+    def test_wrong_keys(self):
+        """
+        Test that instantiating a Rule with mistyped keys fails.
+        """
+        try:
+            Rule(action="deny", dst_nets="192.168.13.0/24")
+        except KeyError:
+            pass
+        else:
+            self.assertTrue(False, "Should raise key error.")
+
+    def test_wrong_value(self):
+        """
+        Test that instantiating a Rule with action not allow|deny will fail.
+        """
+        try:
```

Comment:
More 'noseish' to use self.assert_raises instead of writing the boilerplate to try/catch? I think it's less code & clearer.

There's also the @raises decorator, but I prefer the former option.